bundled/tool/ruff_server.py

```diff
@@ -0,0 +1,89 @@
+import os
+import shutil
+import site
+import subprocess
+import sys
+import sysconfig
+from pathlib import Path
+
+RUFF_EXE = "ruff.exe" if sys.platform == "win32" else "ruff"
+
+BUNDLE_DIR = Path(__file__).parent.parent
+
+
+def update_sys_path(path_to_add: str) -> None:
+    """Add given path to `sys.path`."""
+    if os.path.isdir(path_to_add):
+        # The `site` module adds the directory at the end, if not yet present; we want
+        # it to be at the beginning, so that it takes precedence over any other
+        # installed versions.
+        sys.path.insert(0, path_to_add)
+
+        # Allow development versions of libraries to be imported.
+        site.addsitedir(path_to_add)
+
+
+if __name__ == "__main__":
+    # Ensure that we can import bundled libraries like `packaging`
+    update_sys_path(os.fspath(BUNDLE_DIR / "libs"))
+
+
+from packaging.specifiers import SpecifierSet
+from packaging.version import Version
+
+# This is the first release that included `ruff server`.
+# The minimum version may change in the future.
+RUFF_VERSION_REQUIREMENT = SpecifierSet(">=0.3.3")
+# These versions have major bugs or broken integration, and should be avoided.
+FORBIDDEN_RUFF_VERSIONS = [Version("0.3.4")]
+
+
+def executable_version(executable: str) -> Version:
+    """Return the version of the executable at the given path."""
+    output = subprocess.check_output([executable, "--version"]).decode().strip()
+    version = output.replace("ruff ", "")
+    return Version(version)
+
+
+def check_compatibility(
+    executable: str,
+    requirement: SpecifierSet,
+    forbidden_requirements: SpecifierSet,
+) -> None:
+    """Check the executable for compatibility against various version specifiers."""
+    version = executable_version(executable)
+    if not requirement.contains(version, prereleases=True):
+        message = f"Ruff {requirement} required, but found {version} at {executable}"
+        raise RuntimeError(message)
+    for forbidden in forbidden_requirements:
+        if version == forbidden:
+            message = (
+                f"Tried to use Ruff version {version} at {executable}, which is not allowed.\n"
+                "This version of the server has incompatibilities and/or integration-breaking bugs.\n"
+                "Please upgrade to the latest version and try again."
+            )
+            raise RuntimeError(message)
+
+
+def find_ruff_bin(fallback: Path) -> Path:
+    """Return the ruff binary path."""
+    path = Path(sysconfig.get_path("scripts")) / RUFF_EXE
+    if path.is_file():
+        return path
```

Comment:
I think we might as well change this to use the `find_ruff_bin` logic that we use in Ruff itself:

```python
def find_ruff_bin() -> str:
    """Return the ruff binary path."""

    ruff_exe = "ruff" + sysconfig.get_config_var("EXE")

    path = os.path.join(sysconfig.get_path("scripts"), ruff_exe)
    if os.path.isfile(path):
        return path

    if sys.version_info >= (3, 10):
        user_scheme = sysconfig.get_preferred_scheme("user")
    elif os.name == "nt":
        user_scheme = "nt_user"
    elif sys.platform == "darwin" and sys._framework:
        user_scheme = "osx_framework_user"
    else:
        user_scheme = "posix_user"

    path = os.path.join(sysconfig.get_path("scripts", scheme=user_scheme), ruff_exe)
    if os.path.isfile(path):
        return path

    raise FileNotFoundError(path)
```

(Like, don't call Ruff, just add this function here.)